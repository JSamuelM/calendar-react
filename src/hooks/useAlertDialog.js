import { useDisclosure } from '@chakra-ui/react';
import { useState } from 'react';

export const useAlertDialog = () => {
  const { isOpen, onOpen, onClose, onToggle } = useDisclosure();
  const [alertConfig, setAlertConfig] = useState({
    title: 'title',
    body: 'body',
    action: 'Save',
    colorScheme: 'blue'
  });

  const configAlert = ({ title, body, action, colorScheme }) => {
    setAlertConfig({
      title,
      body,
      action,
      colorScheme,
    });
  }

  return {
    isOpen,
    alertConfig,

    onToggle,
    onOpen,
    onClose,
    configAlert
  }
}
