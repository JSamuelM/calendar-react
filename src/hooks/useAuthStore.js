import { useDispatch, useSelector } from "react-redux"
import calendarApi from "../api/calendarApi";
import { clearErrorMessage, onChecking, onLogin, onLogout, onLogoutCalendar } from "../store";

export const useAuthStore = () => {

  const { status, user, errorMessage } = useSelector(state => state.auth);
  const dispatch = useDispatch();

  const startLogin = async ({ email, password }) => {
    dispatch(onChecking());
    try {
      const { data } = await calendarApi.post('/auth', { email, password });
      localStorage.setItem('token', data.token);
      localStorage.setItem('token-init-date', new Date().getTime());
      dispatch(onLogin({ name: data.name, uid: data.uid }));

    } catch (error) {
      dispatch(onLogout('Credenciales incorrectas'));
    }
  }

  const resetErrorMessage = ({ msg }) => {
    dispatch(clearErrorMessage(msg));
  }

  const startRegister = async ({ name, email, password }) => {
    try {
      const { data } = await calendarApi.post('/auth/new', { name, email, password });
      localStorage.setItem('token', data.token);
      localStorage.setItem('token-init-date', new Date().getTime());
      dispatch(onLogin({ name: data.name, uid: data.uid }));
    } catch (error) {
      const errors = error.response.data?.msg ||
        Object.values(error.response.data?.errors)
          .map((message) => message.msg)
          .join();
      dispatch(onLogout(errors));
    }
  }

  const checkAuthToken = async () => {
    const token = localStorage.getItem('token');
    if (!token) {
      return dispatch(onLogout());
    }

    try {
      const { data } = await calendarApi.get('/auth/renew');
      localStorage.setItem('token', data.token);
      localStorage.setItem('token-init-date', new Date().getTime());
      dispatch(onLogin({ name: data.name, uid: data.uid }));
    } catch (error) {
      localStorage.removeItem('token');
      dispatch(onLogout());
    }
  }

  const startLogout = () => {
    localStorage.removeItem('token');
    localStorage.removeItem('token-init-date');
    localStorage.removeItem('lastView');
    dispatch(onLogout());
    dispatch(onLogoutCalendar());
  }

  return {
    // * Props
    errorMessage,
    status,
    user,

    // * Methods
    checkAuthToken,
    startLogin,
    startLogout,
    startRegister,
    resetErrorMessage
  }
}
