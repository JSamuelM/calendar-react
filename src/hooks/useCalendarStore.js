import { useDispatch, useSelector } from 'react-redux'
import { onAddNewEvent, onDeleteEvent, onLoadEvents, onSetActiveEvent, onUpdateEvent } from '../store';
import calendarApi from '../api/calendarApi';
import { convertEventsToDateEvents } from '../helpers';

export const useCalendarStore = () => {

  const dispatch = useDispatch();
  const { activeEvent, events } = useSelector(state => state.calendar);
  const { user } = useSelector(state => state.auth);

  const setActiveEvent = (calendarEvent) => {
    dispatch(onSetActiveEvent(calendarEvent));
  }

  const startSavingEvent = async (calendarEvent) => {
    try {
      if (calendarEvent.id) {
        // updating
        await calendarApi.put(`/events/${calendarEvent.id}`, calendarEvent);

        dispatch(onUpdateEvent({ ...calendarEvent, user }));
        return;
      }

      // creating
      const { data } = await calendarApi.post('/events', calendarEvent);
      dispatch(onAddNewEvent({ ...calendarEvent, id: data.event.id, user }));
    } catch (error) {
      console.error(error);
      return error.response.data;
    }
  }

  const startDeletingEvent = async () => {
    try {
      await calendarApi.delete(`/events/${activeEvent.id}`)
      dispatch(onDeleteEvent());
    } catch (error) {
      console.error('Error: ' + error.response);
      return error.response.data;
    }
  }

  const startLoadingEvents = async () => {
    try {
      const { data } = await calendarApi.get('/events');
      const events = convertEventsToDateEvents(data.events)
      dispatch(onLoadEvents(events));
    } catch (error) {
      console.log('Error');
      console.log(error);
    }
  }

  return {
    // Propierties
    activeEvent,
    events,
    hasEventSelected: !!activeEvent?.id,

    // Methods
    setActiveEvent,
    startDeletingEvent,
    startLoadingEvents,
    startSavingEvent,
  }
}