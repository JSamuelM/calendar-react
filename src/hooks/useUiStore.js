import { useDispatch, useSelector } from 'react-redux'
import { onAlertClose, onAlertOpen, onCloseDateModal, onOpenDateModal } from '../store';

export const useUiStore = () => {

  const dispatch = useDispatch();

  const {
    isAlertOpen,
    isDateModalOpen
  } = useSelector(state => state.ui);

  const openDateModal = () => {
    dispatch(onOpenDateModal());
  }

  const closeDateModal = () => {
    dispatch(onCloseDateModal());
  }

  const openAlert = () => {
    dispatch(onAlertOpen());
  }

  const closeAlert = () => {
    dispatch(onAlertClose());
  }

  const toggleDateModal = () => {
    isDateModalOpen ? openDateModal() : closeDateModal();
  }

  return {
    // Propierties
    isDateModalOpen,
    isAlertOpen,

    // Methods
    closeDateModal,
    openDateModal,
    toggleDateModal,
    openAlert,
    closeAlert,
  }
}
