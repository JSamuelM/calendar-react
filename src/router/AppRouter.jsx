import { Navigate, Route, Routes } from 'react-router-dom';
import { LoginPage } from '../auth';
import { CalendarPage } from '../calendar';
import { useAuthStore } from '../hooks';
import { useEffect } from 'react';
import { Center, Flex, Spinner, Text, VStack } from '@chakra-ui/react';

export const AppRouter = () => {

  const { status, checkAuthToken } = useAuthStore();
  // const authStatus = 'not-authenticated'; // 'authenticated'

  useEffect(() => {
    checkAuthToken()
  }, []);

  if (status === 'checking') {
    return (
      <Flex height="100vh" justifyContent="center">
        <Center>
          <VStack spacing={ 4 }>
            <Spinner size="xl" />
            <Text fontSize="xl">Loading...</Text>
          </VStack>
        </Center>
      </Flex>
    )
  }  

  return (
    <Routes>
      {
        (status === 'authenticated')
          ? (
            <>
              <Route path="/" element={ <CalendarPage /> } />
              <Route path="/*" element={ <Navigate to="/" /> } />
            </>
          )
          : (
            <>
              <Route path="/auth/*" element={ <LoginPage /> } />
              <Route path="/*" element={ <Navigate to="/auth/login" /> } />
            </>
          )
      }
    </Routes>
  )
}
