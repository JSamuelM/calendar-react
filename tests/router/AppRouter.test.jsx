import { render, screen } from '@testing-library/react';
import { useAuthStore } from '../../src/hooks/useAuthStore';
import { AppRouter } from '../../src/router/AppRouter';
import { MemoryRouter } from 'react-router-dom';

jest.mock('../../src/hooks/useAuthStore');

jest.mock('../../src/calendar', () => ({
  CalendarPage: () => <h1>CalendarPage</h1>
}));

describe('Pruebas en <AppRouter />', () => {

  const mockCheckAuthToken = jest.fn();

  beforeEach(() => jest.clearAllMocks());

  test('debe de mostrar la pantall de carga y llamar checkAuthToken', () => {
    
    useAuthStore.mockReturnValue({
      status: 'checking',
      checkAuthToken: mockCheckAuthToken
    });

    render(<AppRouter />);

    expect(screen.getAllByText('Loading...')).toBeTruthy();
    expect(mockCheckAuthToken).toHaveBeenCalled();
  });

  test('debe de mostrar el login en caso de no estar autenticado', () => {
    useAuthStore.mockReturnValue({
      status: 'not-authenticated',
      checkAuthToken: mockCheckAuthToken
    });

    const { container } = render(
      <MemoryRouter>
        <AppRouter />
      </MemoryRouter>
    );

    expect(screen.getAllByText('Sign up')).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  test('debe de mostrar el calendario si estamos autenticados', () => {
    useAuthStore.mockReturnValue({
      status: 'authenticated',
      checkAuthToken: mockCheckAuthToken
    });

    render(
      <MemoryRouter>
        <AppRouter />
      </MemoryRouter>
    );

    
  });


});
