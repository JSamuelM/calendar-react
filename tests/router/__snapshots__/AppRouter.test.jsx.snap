// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <AppRouter /> debe de mostrar el login en caso de no estar autenticado 1`] = `
<div>
  <div
    class="css-1jc7f17"
  >
    <div
      class="css-2j9zgb"
    >
      <button
        class="chakra-button css-4xx2wk"
        type="button"
      >
        <svg
          class="chakra-icon css-onkibi"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <g
            fill="none"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          >
            <circle
              cx="12"
              cy="12"
              r="5"
            />
            <path
              d="M12 1v2"
            />
            <path
              d="M12 21v2"
            />
            <path
              d="M4.22 4.22l1.42 1.42"
            />
            <path
              d="M18.36 18.36l1.42 1.42"
            />
            <path
              d="M1 12h2"
            />
            <path
              d="M21 12h2"
            />
            <path
              d="M4.22 19.78l1.42-1.42"
            />
            <path
              d="M18.36 5.64l1.42-1.42"
            />
          </g>
        </svg>
      </button>
    </div>
    <div
      class="chakra-stack css-zgufkz"
    >
      <div
        class="chakra-stack css-tl3ftk"
      >
        <h2
          class="chakra-heading css-5efzzv"
        >
          Sign up
        </h2>
        <p
          class="chakra-text css-1bv0cc9"
        >
          to enjoy all of our cool features ✌️
        </p>
      </div>
      <div
        class="css-3hziw7"
      >
        <div
          class="chakra-tabs css-79elbk"
        >
          <div
            aria-orientation="horizontal"
            class="chakra-tabs__tablist css-k008qs"
            role="tablist"
          >
            <button
              aria-controls="tabs-:r2:--tabpanel-0"
              aria-disabled="false"
              aria-selected="true"
              class="chakra-tabs__tab css-5wl5ud"
              data-index="0"
              id="tabs-:r2:--tab-0"
              role="tab"
              tabindex="0"
              type="button"
            >
              Sign up
            </button>
            <button
              aria-controls="tabs-:r2:--tabpanel-1"
              aria-disabled="false"
              aria-selected="false"
              class="chakra-tabs__tab css-5wl5ud"
              data-index="1"
              id="tabs-:r2:--tab-1"
              role="tab"
              tabindex="-1"
              type="button"
            >
              Register
            </button>
          </div>
          <div
            class="chakra-tabs__tab-panels css-8atqhb"
          >
            <div
              aria-labelledby="tabs-:r2:--tab-0"
              class="chakra-tabs__tab-panel css-sjt5nk"
              id="tabs-:r2:--tabpanel-0"
              role="tabpanel"
              tabindex="0"
            >
              <div
                class="chakra-stack css-1kkt86i"
              >
                <form>
                  <div
                    class="chakra-form-control css-0"
                    role="group"
                  >
                    <label
                      class="chakra-form__label css-uvho5g"
                      for="loginEmail"
                      id="loginEmail-label"
                    >
                      Email address
                      <span
                        aria-hidden="true"
                        class="chakra-form__required-indicator css-0"
                        role="presentation"
                      >
                        *
                      </span>
                    </label>
                    <input
                      aria-required="true"
                      class="chakra-input css-0"
                      id="loginEmail"
                      name="loginEmail"
                      required=""
                      type="email"
                      value=""
                    />
                  </div>
                  <div
                    class="chakra-form-control css-0"
                    role="group"
                  >
                    <label
                      class="chakra-form__label css-uvho5g"
                      for="loginPassword"
                      id="loginPassword-label"
                    >
                      Password
                      <span
                        aria-hidden="true"
                        class="chakra-form__required-indicator css-0"
                        role="presentation"
                      >
                        *
                      </span>
                    </label>
                    <div
                      class="chakra-input__group css-bx0blc"
                      data-group="true"
                    >
                      <input
                        aria-required="true"
                        class="chakra-input css-0"
                        id="loginPassword"
                        name="loginPassword"
                        required=""
                        type="password"
                        value=""
                      />
                      <div
                        class="chakra-input__right-element css-1aezdzf"
                      >
                        <button
                          class="chakra-button css-4xx2wk"
                          type="button"
                        >
                          <svg
                            class="chakra-icon css-onkibi"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <g
                              fill="currentColor"
                            >
                              <path
                                d="M23.2,10.549a20.954,20.954,0,0,0-4.3-3.6l4-3.995a1,1,0,1,0-1.414-1.414l-.018.018a.737.737,0,0,1-.173.291l-19.5,19.5c-.008.007-.018.009-.026.017a1,1,0,0,0,1.631,1.088l4.146-4.146a11.26,11.26,0,0,0,4.31.939h.3c4.256,0,8.489-2.984,11.051-5.8A2.171,2.171,0,0,0,23.2,10.549ZM16.313,13.27a4.581,4.581,0,0,1-3,3.028,4.3,4.3,0,0,1-3.1-.19.253.253,0,0,1-.068-.407l5.56-5.559a.252.252,0,0,1,.407.067A4.3,4.3,0,0,1,16.313,13.27Z"
                              />
                              <path
                                d="M7.615,13.4a.244.244,0,0,0,.061-.24A4.315,4.315,0,0,1,7.5,12,4.5,4.5,0,0,1,12,7.5a4.276,4.276,0,0,1,1.16.173.244.244,0,0,0,.24-.062l1.941-1.942a.254.254,0,0,0-.1-.421A10.413,10.413,0,0,0,12,4.75C7.7,4.692,3.4,7.7.813,10.549a2.15,2.15,0,0,0-.007,2.9,21.209,21.209,0,0,0,3.438,3.03.256.256,0,0,0,.326-.029Z"
                              />
                            </g>
                          </svg>
                        </button>
                      </div>
                    </div>
                  </div>
                  <div
                    class="chakra-stack css-1fivd04"
                  >
                    <button
                      class="chakra-button css-t3msxv"
                      type="submit"
                    >
                      Sign up
                    </button>
                  </div>
                </form>
              </div>
            </div>
            <div
              aria-labelledby="tabs-:r2:--tab-1"
              class="chakra-tabs__tab-panel css-sjt5nk"
              hidden=""
              id="tabs-:r2:--tabpanel-1"
              role="tabpanel"
              tabindex="0"
            >
              <div
                class="chakra-stack css-1kkt86i"
              >
                <form>
                  <div
                    class="chakra-form-control css-0"
                    role="group"
                  >
                    <label
                      class="chakra-form__label css-uvho5g"
                      for="registerName"
                      id="registerName-label"
                    >
                      Name
                      <span
                        aria-hidden="true"
                        class="chakra-form__required-indicator css-0"
                        role="presentation"
                      >
                        *
                      </span>
                    </label>
                    <input
                      aria-required="true"
                      class="chakra-input css-0"
                      id="registerName"
                      name="registerName"
                      required=""
                      type="text"
                      value=""
                    />
                  </div>
                  <div
                    class="chakra-form-control css-0"
                    role="group"
                  >
                    <label
                      class="chakra-form__label css-uvho5g"
                      for="registerEmail"
                      id="registerEmail-label"
                    >
                      Email address
                      <span
                        aria-hidden="true"
                        class="chakra-form__required-indicator css-0"
                        role="presentation"
                      >
                        *
                      </span>
                    </label>
                    <input
                      aria-required="true"
                      class="chakra-input css-0"
                      id="registerEmail"
                      name="registerEmail"
                      required=""
                      type="email"
                      value=""
                    />
                  </div>
                  <div
                    class="chakra-form-control css-0"
                    role="group"
                  >
                    <label
                      class="chakra-form__label css-uvho5g"
                      for="registerPassword"
                      id="registerPassword-label"
                    >
                      Password
                      <span
                        aria-hidden="true"
                        class="chakra-form__required-indicator css-0"
                        role="presentation"
                      >
                        *
                      </span>
                    </label>
                    <div
                      class="chakra-input__group css-bx0blc"
                      data-group="true"
                    >
                      <input
                        aria-required="true"
                        class="chakra-input css-0"
                        id="registerPassword"
                        name="registerPassword"
                        required=""
                        type="password"
                        value=""
                      />
                      <div
                        class="chakra-input__right-element css-1aezdzf"
                      >
                        <button
                          class="chakra-button css-4xx2wk"
                          type="button"
                        >
                          <svg
                            class="chakra-icon css-onkibi"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <g
                              fill="currentColor"
                            >
                              <path
                                d="M23.2,10.549a20.954,20.954,0,0,0-4.3-3.6l4-3.995a1,1,0,1,0-1.414-1.414l-.018.018a.737.737,0,0,1-.173.291l-19.5,19.5c-.008.007-.018.009-.026.017a1,1,0,0,0,1.631,1.088l4.146-4.146a11.26,11.26,0,0,0,4.31.939h.3c4.256,0,8.489-2.984,11.051-5.8A2.171,2.171,0,0,0,23.2,10.549ZM16.313,13.27a4.581,4.581,0,0,1-3,3.028,4.3,4.3,0,0,1-3.1-.19.253.253,0,0,1-.068-.407l5.56-5.559a.252.252,0,0,1,.407.067A4.3,4.3,0,0,1,16.313,13.27Z"
                              />
                              <path
                                d="M7.615,13.4a.244.244,0,0,0,.061-.24A4.315,4.315,0,0,1,7.5,12,4.5,4.5,0,0,1,12,7.5a4.276,4.276,0,0,1,1.16.173.244.244,0,0,0,.24-.062l1.941-1.942a.254.254,0,0,0-.1-.421A10.413,10.413,0,0,0,12,4.75C7.7,4.692,3.4,7.7.813,10.549a2.15,2.15,0,0,0-.007,2.9,21.209,21.209,0,0,0,3.438,3.03.256.256,0,0,0,.326-.029Z"
                              />
                            </g>
                          </svg>
                        </button>
                      </div>
                    </div>
                  </div>
                  <div
                    class="chakra-form-control css-0"
                    role="group"
                  >
                    <label
                      class="chakra-form__label css-uvho5g"
                      for="confirmPassword"
                      id="confirmPassword-label"
                    >
                      Confirm password
                      <span
                        aria-hidden="true"
                        class="chakra-form__required-indicator css-0"
                        role="presentation"
                      >
                        *
                      </span>
                    </label>
                    <div
                      class="chakra-input__group css-bx0blc"
                      data-group="true"
                    >
                      <input
                        aria-required="true"
                        class="chakra-input css-0"
                        id="confirmPassword"
                        name="registerPassword2"
                        required=""
                        type="password"
                        value=""
                      />
                      <div
                        class="chakra-input__right-element css-1aezdzf"
                      >
                        <button
                          class="chakra-button css-4xx2wk"
                          type="button"
                        >
                          <svg
                            class="chakra-icon css-onkibi"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <g
                              fill="currentColor"
                            >
                              <path
                                d="M23.2,10.549a20.954,20.954,0,0,0-4.3-3.6l4-3.995a1,1,0,1,0-1.414-1.414l-.018.018a.737.737,0,0,1-.173.291l-19.5,19.5c-.008.007-.018.009-.026.017a1,1,0,0,0,1.631,1.088l4.146-4.146a11.26,11.26,0,0,0,4.31.939h.3c4.256,0,8.489-2.984,11.051-5.8A2.171,2.171,0,0,0,23.2,10.549ZM16.313,13.27a4.581,4.581,0,0,1-3,3.028,4.3,4.3,0,0,1-3.1-.19.253.253,0,0,1-.068-.407l5.56-5.559a.252.252,0,0,1,.407.067A4.3,4.3,0,0,1,16.313,13.27Z"
                              />
                              <path
                                d="M7.615,13.4a.244.244,0,0,0,.061-.24A4.315,4.315,0,0,1,7.5,12,4.5,4.5,0,0,1,12,7.5a4.276,4.276,0,0,1,1.16.173.244.244,0,0,0,.24-.062l1.941-1.942a.254.254,0,0,0-.1-.421A10.413,10.413,0,0,0,12,4.75C7.7,4.692,3.4,7.7.813,10.549a2.15,2.15,0,0,0-.007,2.9,21.209,21.209,0,0,0,3.438,3.03.256.256,0,0,0,.326-.029Z"
                              />
                            </g>
                          </svg>
                        </button>
                      </div>
                    </div>
                  </div>
                  <div
                    class="chakra-stack css-1fivd04"
                  >
                    <button
                      class="chakra-button css-t3msxv"
                      type="submit"
                    >
                      Register
                    </button>
                  </div>
                </form>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
